File: web/src/main/java/no/nav/sbl/dialogarena/rest/ressurser/bostotte/BostotteRessurs.java
package no.nav.sbl.dialogarena.rest.ressurser.bostotte;

import no.nav.metrics.aspects.Timed;
import no.nav.sbl.dialogarena.sendsoknad.domain.oidc.OidcFeatureToggleUtils;
import no.nav.sbl.soknadsosialhjelp.digisos.soker.hendelse.JsonSaksStatus;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomiOpplysningUtbetaling;
import no.nav.sbl.sosialhjelp.soknadunderbehandling.SoknadUnderArbeidRepository;
import no.nav.security.oidc.api.ProtectedWithClaims;
import org.springframework.stereotype.Controller;

import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import java.util.List;
import java.util.stream.Collectors;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
import static no.nav.sbl.dialogarena.bostotte.Bostotte.HUSBANKEN_TYPE;

@Controller
@ProtectedWithClaims(issuer = "selvbetjening", claimMap = { "acr=Level4" })
@Path("/soknader/{behandlingsId}/bostotte")
@Timed
@Produces(APPLICATION_JSON)
public class BostotteRessurs {

    @Inject
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @GET
    public BostotteFrontend hentBostotte(@PathParam("behandlingsId") String behandlingsId) {
        final String eier = OidcFeatureToggleUtils.getUserId();
        final JsonInternalSoknad soknad = soknadUnderArbeidRepository.hentSoknad(behandlingsId, eier).getJsonInternalSoknad();

        return new BostotteFrontend()
                .withUtbetalinger(mapToUtbetalinger(soknad))
                .withSaksStatuser(mapToUtSaksStatuser(soknad));
    }

    private List<JsonOkonomiOpplysningUtbetaling> mapToUtbetalinger(JsonInternalSoknad soknad) {
        return soknad.getSoknad().getData().getOkonomi().getOpplysninger().getUtbetaling().stream()
                .filter(utbetaling -> utbetaling.getType().equals(HUSBANKEN_TYPE))
                .collect(Collectors.toList());
    }

    private List<JsonSaksStatus> mapToUtSaksStatuser(JsonInternalSoknad soknad) {
        return null;
    }

    @XmlAccessorType(XmlAccessType.FIELD)
    public static final class BostotteFrontend {
        public List<JsonOkonomiOpplysningUtbetaling> utbetalinger;
        public List<JsonSaksStatus> saksStatuser;

        public BostotteFrontend withUtbetalinger(List<JsonOkonomiOpplysningUtbetaling> utbetalinger) {
            this.utbetalinger = utbetalinger;
            return this;
        }

        public BostotteFrontend withSaksStatuser(List<JsonSaksStatus> saksStatuser) {
            this.saksStatuser = saksStatuser;
            return this;
        }
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/inntekt/husbanken/BostotteRessurs.kt
package no.nav.sosialhjelp.soknad.inntekt.husbanken

import no.nav.sbl.soknadsosialhjelp.soknad.bostotte.JsonBostotteSak
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomiOpplysningUtbetaling
import no.nav.security.token.support.core.api.ProtectedWithClaims
import no.nav.sosialhjelp.soknad.app.Constants
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@ProtectedWithClaims(
    issuer = Constants.SELVBETJENING,
    claimMap = [Constants.CLAIM_ACR_LEVEL_4, Constants.CLAIM_ACR_LOA_HIGH],
    combineWithOr = true,
)
@RequestMapping("/soknader/{behandlingsId}/inntekt/bostotte", produces = [MediaType.APPLICATION_JSON_VALUE])
class BostotteRessurs(
    private val tilgangskontroll: Tilgangskontroll,
    private val bostotteProxy: BostotteProxy,
) {
    @GetMapping
    fun hentBostotte(
        @PathVariable("behandlingsId") behandlingsId: String,
    ): BostotteFrontend {
        tilgangskontroll.verifiserAtBrukerHarTilgang()

        return bostotteProxy.getBostotte(behandlingsId)
    }

    @PutMapping
    fun updateBostotte(
        @PathVariable("behandlingsId") behandlingsId: String,
        @RequestBody bostotteFrontend: BostotteFrontend,
    ) {
        tilgangskontroll.verifiserAtBrukerKanEndreSoknad(behandlingsId)

        bostotteProxy.updateBostotteBekreftelse(
            soknadId = behandlingsId,
            hasBostotte = bostotteFrontend.bekreftelse,
        )
    }

    @PostMapping("/samtykke")
    fun updateSamtykke(
        @PathVariable("behandlingsId") behandlingsId: String,
        @RequestBody samtykke: Boolean,
        @RequestHeader(value = HttpHeaders.AUTHORIZATION) token: String?,
    ) {
        tilgangskontroll.verifiserAtBrukerKanEndreSoknad(behandlingsId)

        bostotteProxy.updateBostotteSamtykke(behandlingsId, samtykke, token)
    }

    data class BostotteFrontend(
        val bekreftelse: Boolean?,
        val samtykke: Boolean?,
        val utbetalinger: List<JsonOkonomiOpplysningUtbetaling>?,
        val saker: List<JsonBostotteSak>?,
        val stotteFraHusbankenFeilet: Boolean?,
        val samtykkeTidspunkt: String?,
    )
}
